src/server/handlers/dynamic.js

```diff
@@ -24,7 +25,26 @@ export default ({ server, config }) => {
       // Set a cache key
       const currentPath = request.url.pathname || '/'
       const isPreview = Boolean(request.query && request.query.tapestry_hash)
-      const cacheKey = normaliseUrlPath(currentPath)
+      let cacheKey = normaliseUrlPath(currentPath)
+
+      // Run cacheKeyHandler function if provided by client
+      const cacheKeyHandler = config.cacheKeyHandler
+      let newCacheKey = cacheKey
+      if (isFunction(cacheKeyHandler)) {
+        try {
+          newCacheKey = cacheKeyHandler(request, cacheKey)
+          if (typeof newCacheKey === 'string') {
+            cacheKey = newCacheKey
+          } else {
+            console.error(
+              `cacheKeyHandler() return value: expected string but received ${typeof cacheKey}`
+            )
+          }
+        } catch (e) {
+          console.error(`cacheKeyHandler() error: ${e}`)
+        }
```

Comment:
We should be able to `throw` within a `try`, e.g. `if (typeof newCacheKey !== 'string') throw 'Error text'`. That can then be handled by the `catch` in the same way.

I reckon we should utilise our logger for errors, you should have `log.error()` available within `utilities/logger`

Reply:
Very nice and useful! Thanks 🙏 😊